src/main/java/org/apache/maven/shared/artifact/filter/PatternIncludesArtifactFilter.java

```diff
@@ -107,322 +147,783 @@ public boolean include( final Artifact artifact )
      */
     protected boolean patternMatches( final Artifact artifact )
     {
-        return positiveMatch( artifact ) == Boolean.TRUE || negativeMatch( artifact ) == Boolean.FALSE;
+        // Check if the main artifact matches
+        char[][] tokens = new char[][] {
```

Comment:
`tokens` doesn't describe this clearly, espacially if the order matters. Something like `gatvAsCharArray` would at least help to understand how it is used later on on the code. (sometimes type and version are switched, so better to add that explcitly in the variable name)

Reply:
Right, I'll also push a few more comments.